File: leetcode solutions/Python/Happy Number/happyNumber.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """

        mem = set()
        while n != 1:
            n = sum([int(i)**2 for i in str(n)])
            if n not in mem:
                mem.add(n)
            else:
                return False
        return True



# solution 2 : more elegant but TLE

'''
class Solution:

    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        lookup = {}
        while n != 1 or n not in lookup:
            lookup[n] = True
            n = self.nextNumber(n)
        return n == 1


    def nextNumber(self, n):
        new = 0
        for ch in str(n):
            new += int(ch) ** 2
        return new
'''

File: leetcode solutions/C++/Happy Number/happyNumber.cpp
// O(1) space

class Solution {
public:

    int digitSquareSum(int n) {
        int sum = 0, tmp;
        while (n != 0) {
            tmp = n % 10;
            sum += tmp * tmp;
            n /= 10;
        }
        return sum;
    }

    bool isHappy(int n) {
        int slow = n, fast = digitSquareSum(n);

        while (slow != fast) {
            slow = digitSquareSum(slow);
            fast = digitSquareSum(digitSquareSum(fast));
        }

        return slow == 1;

    }

};